File: apps/mac/Sources/T3Kit/ParentThinking.swift
import Foundation

/// Signals used to decide whether the parent agent should show an ephemeral
/// "Thinking…" indicator. Mirrors `packages/shared/src/parentThinking.ts`.
public struct ParentThinkingSignals: Sendable, Equatable {
    /// Orchestration session status (`running`, `waiting`, `idle`, …).
    public var sessionStatus: String?
    /// Latest turn state (`running`, `completed`, `error`, `interrupted`, …).
    public var latestTurnState: String?
    public var hasPendingApproval: Bool
    public var hasPendingUserInput: Bool
    /// Server-authoritative stall from `session.health`.
    public var isStalled: Bool
    /// Parent-thread tool call currently in progress.
    public var hasActiveToolActivity: Bool
    /// An assistant message is streaming (empty or not).
    public var hasActiveStreamingAssistant: Bool
    /// Non-empty streamed assistant body for the active turn.
    public var hasStreamingAssistantText: Bool
    /// Visible reasoning/progress text for the parent agent.
    public var hasVisibleReasoningText: Bool

    public init(
        sessionStatus: String? = nil,
        latestTurnState: String? = nil,
        hasPendingApproval: Bool = false,
        hasPendingUserInput: Bool = false,
        isStalled: Bool = false,
        hasActiveToolActivity: Bool = false,
        hasActiveStreamingAssistant: Bool = false,
        hasStreamingAssistantText: Bool = false,
        hasVisibleReasoningText: Bool = false
    ) {
        self.sessionStatus = sessionStatus
        self.latestTurnState = latestTurnState
        self.hasPendingApproval = hasPendingApproval
        self.hasPendingUserInput = hasPendingUserInput
        self.isStalled = isStalled
        self.hasActiveToolActivity = hasActiveToolActivity
        self.hasActiveStreamingAssistant = hasActiveStreamingAssistant
        self.hasStreamingAssistantText = hasStreamingAssistantText
        self.hasVisibleReasoningText = hasVisibleReasoningText
    }
}

/// Pure gate for the parent "Thinking…" indicator.
public enum ParentThinking {
    /// True when the parent model is actively reasoning without other visible
    /// activity the UI already surfaces.
    public static func shouldShow(_ signals: ParentThinkingSignals) -> Bool {
        if signals.hasPendingApproval || signals.hasPendingUserInput {
            return false
        }
        if signals.isStalled {
            return false
        }

        let sessionStatus = signals.sessionStatus
        let latestTurnState = signals.latestTurnState

        // Waiting covers sleep/wake and provider compaction (Claude reports
        // compacting as session waiting). Error/interrupted/stopped are terminal.
        switch sessionStatus {
        case "waiting", "error", "interrupted", "stopped", "idle", "ready":
            return false
        default:
            break
        }

        switch latestTurnState {
        case "error", "interrupted", "completed":
            if sessionStatus != "running", sessionStatus != "starting" {
                return false
            }
        default:
            break
        }

        let turnActive =
            sessionStatus == "running"
            || sessionStatus == "starting"
            || latestTurnState == "running"
        guard turnActive else { return false }

        if signals.hasActiveToolActivity { return false }
        if signals.hasActiveStreamingAssistant || signals.hasStreamingAssistantText {
            return false
        }
        if signals.hasVisibleReasoningText { return false }

        return true
    }
}

File: packages/shared/src/parentThinking.ts
/**
 * Whether the parent agent should show an ephemeral "Thinking…" indicator.
 *
 * Authoritative runtime state is the only entry gate: the session/turn must
 * report active work. Visible model output (streamed assistant text), tool
 * execution, waiting/compaction, approvals, stalls, and terminal outcomes all
 * suppress the indicator so it never competes with a more specific state and
 * never floods the transcript with repeated entries.
 */

export type ParentThinkingSignals = {
  /** Orchestration session status (`running`, `waiting`, `idle`, …). */
  readonly sessionStatus: string | null | undefined;
  /** Latest turn state (`running`, `completed`, `error`, `interrupted`, …). */
  readonly latestTurnState: string | null | undefined;
  readonly hasPendingApproval?: boolean;
  readonly hasPendingUserInput?: boolean;
  /** Server-authoritative stall from `session.health`. */
  readonly isStalled?: boolean;
  /** Parent-thread tool call currently in progress. */
  readonly hasActiveToolActivity?: boolean;
  /**
   * An assistant message is streaming (empty or not). The stream row itself
   * communicates progress, including the empty-placeholder case.
   */
  readonly hasActiveStreamingAssistant?: boolean;
  /** Non-empty streamed assistant body for the active turn. */
  readonly hasStreamingAssistantText?: boolean;
  /** Visible reasoning/progress text for the parent agent. */
  readonly hasVisibleReasoningText?: boolean;
};

/**
 * True when the parent model is actively reasoning without other visible
 * activity the UI already surfaces.
 */
export function shouldShowParentThinking(signals: ParentThinkingSignals): boolean {
  if (signals.hasPendingApproval || signals.hasPendingUserInput) {
    return false;
  }
  if (signals.isStalled) {
    return false;
  }

  const sessionStatus = signals.sessionStatus ?? null;
  const latestTurnState = signals.latestTurnState ?? null;

  // Waiting covers sleep/wake and provider compaction (Claude reports
  // compacting as session waiting). Error/interrupted/stopped are terminal.
  if (
    sessionStatus === "waiting" ||
    sessionStatus === "error" ||
    sessionStatus === "interrupted" ||
    sessionStatus === "stopped" ||
    sessionStatus === "idle" ||
    sessionStatus === "ready"
  ) {
    return false;
  }
  if (
    latestTurnState === "error" ||
    latestTurnState === "interrupted" ||
    latestTurnState === "completed"
  ) {
    // Still allow when the session itself remains running (e.g. turn flipped
    // while the next turn is spinning up), but only via sessionStatus below.
    if (sessionStatus !== "running" && sessionStatus !== "starting") {
      return false;
    }
  }

  const turnActive =
    sessionStatus === "running" || sessionStatus === "starting" || latestTurnState === "running";
  if (!turnActive) {
    return false;
  }

  if (signals.hasActiveToolActivity) {
    return false;
  }
  if (signals.hasActiveStreamingAssistant || signals.hasStreamingAssistantText) {
    return false;
  }
  if (signals.hasVisibleReasoningText) {
    return false;
  }

  return true;
}
